File: functions/src/lib/platforms/publishers/twitter.ts
/**
 * Twitter/X API v2 publisher — real implementation.
 * Port of apps/api/app/services/publisher.py TwitterPublisher.
 */

import type { PlatformPublisher, PublishResult, DecryptedTokens, RefreshedTokens } from "./base.js";
import { TWITTER_CLIENT_ID } from "../../../config/env.js";

const TWEETS_URL = "https://api.twitter.com/2/tweets";
const TOKEN_URL = "https://api.twitter.com/2/oauth2/token";

export class TwitterPublisher implements PlatformPublisher {
  async publish(content: string, metadata: Record<string, unknown>, tokens: DecryptedTokens): Promise<PublishResult> {
    const formatType = (metadata.format_type as string) || "twitter_single";

    try {
      if (formatType === "twitter_thread") {
        return await this.publishThread(content, tokens.accessToken);
      }
      return await this.publishSingle(content, tokens.accessToken);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return { success: false, postId: null, url: null, error: `Twitter publish error: ${message}` };
    }
  }

  private async publishSingle(content: string, accessToken: string): Promise<PublishResult> {
    const resp = await fetch(TWEETS_URL, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ text: content }),
    });

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`Twitter API error ${resp.status}: ${body}`);
    }

    const data = await resp.json() as Record<string, unknown>;
    const tweetId = (data.data as Record<string, unknown>).id as string;
    return {
      success: true,
      postId: tweetId,
      url: `https://x.com/i/status/${tweetId}`,
      error: null,
    };
  }

  private async publishThread(content: string, accessToken: string): Promise<PublishResult> {
    // Split by numbered markers or --- separators
    const parts = content
      .split(/\n\s*---\s*\n|\n\s*\d+\/\s*\n?/)
      .map((p) => p.trim())
      .filter((p) => p.length > 0);

    if (parts.length < 2) {
      return await this.publishSingle(content, accessToken);
    }

    let firstTweetId: string | null = null;
    let previousTweetId: string | null = null;

    for (let i = 0; i < parts.length; i++) {
      const payload: Record<string, unknown> = { text: parts[i] };
      if (previousTweetId) {
        payload.reply = { in_reply_to_tweet_id: previousTweetId };
      }

      const resp = await fetch(TWEETS_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!resp.ok) {
        const body = await resp.text();
        throw new Error(`Twitter API error ${resp.status}: ${body}`);
      }

      const data = await resp.json() as Record<string, unknown>;
      const tweetId = (data.data as Record<string, unknown>).id as string;

      if (i === 0) firstTweetId = tweetId;
      previousTweetId = tweetId;

      // Small delay between thread tweets
      if (i < parts.length - 1) {
        await new Promise((resolve) => setTimeout(resolve, 1000));
      }
    }

    return {
      success: true,
      postId: firstTweetId,
      url: `https://x.com/i/status/${firstTweetId}`,
      error: null,
    };
  }

  async validateConnection(_tokens: DecryptedTokens, expiresAt: Date | null): Promise<boolean> {
    if (expiresAt) {
      return expiresAt > new Date();
    }
    return true;
  }

  async refreshToken(tokens: DecryptedTokens): Promise<RefreshedTokens | null> {
    if (!tokens.refreshToken) return null;

    const resp = await fetch(TOKEN_URL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: tokens.refreshToken,
        client_id: TWITTER_CLIENT_ID.value(),
      }),
    });

    if (!resp.ok) return null;

    const data = await resp.json() as Record<string, unknown>;
    return {
      accessToken: data.access_token as string,
      refreshToken: (data.refresh_token as string) ?? tokens.refreshToken,
      expiresIn: (data.expires_in as number) ?? 7200,
    };
  }
}

File: apps/api/app/services/publishers/twitter.py
"""Twitter/X publisher — single tweets and threads."""

import asyncio
import logging
from datetime import datetime, timezone

import httpx

from app.services.publishers.base import BasePublisher, PublishResult
from app.utils.exceptions import ValidationError

logger = logging.getLogger(__name__)

TWEET_MAX_LENGTH = 280
THREAD_DELAY_MS = 500
API_BASE = "https://api.twitter.com/2"


class TwitterPublisher(BasePublisher):
    platform_id = "twitter"

    def validate(self, output: dict) -> None:
        super().validate(output)
        content = output.get("content", "")
        # For threads, content is split by \n---\n
        parts = self._split_thread(content)
        for i, part in enumerate(parts):
            if len(part) > TWEET_MAX_LENGTH:
                raise ValidationError(
                    message="Tweet too long",
                    detail=f"Tweet {i + 1} is {len(part)} chars (max {TWEET_MAX_LENGTH}).",
                )

    async def publish(self, output: dict, token: str, **kwargs) -> PublishResult:
        content = output.get("content", "")
        parts = self._split_thread(content)

        if len(parts) == 1:
            return await self._publish_single(parts[0], token)
        else:
            return await self._publish_thread(parts, token)

    async def _publish_single(self, text: str, token: str) -> PublishResult:
        """Publish a single tweet."""
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(
                f"{API_BASE}/tweets",
                json={"text": text},
                headers={"Authorization": f"Bearer {token}"},
            )

        if resp.status_code not in (200, 201):
            logger.error("Twitter publish failed: %s %s", resp.status_code, resp.text)
            return PublishResult(
                success=False,
                platform="twitter",
                error=f"Twitter API error: {resp.status_code}",
            )

        data = resp.json().get("data", {})
        tweet_id = data.get("id", "")
        return PublishResult(
            success=True,
            platform="twitter",
            platform_post_id=tweet_id,
            platform_post_url=f"https://twitter.com/i/status/{tweet_id}",
            published_at=datetime.now(timezone.utc),
        )

    async def _publish_thread(self, parts: list[str], token: str) -> PublishResult:
        """Publish a thread with 500ms delay between tweets."""
        first_tweet_id = None
        previous_id = None

        for i, text in enumerate(parts):
            payload: dict = {"text": text}
            if previous_id:
                payload["reply"] = {"in_reply_to_tweet_id": previous_id}

            async with httpx.AsyncClient(timeout=15) as client:
                resp = await client.post(
                    f"{API_BASE}/tweets",
                    json=payload,
                    headers={"Authorization": f"Bearer {token}"},
                )

            if resp.status_code not in (200, 201):
                logger.error("Twitter thread part %d failed: %s", i + 1, resp.text)
                return PublishResult(
                    success=False,
                    platform="twitter",
                    platform_post_id=first_tweet_id,
                    error=f"Thread failed at tweet {i + 1}: {resp.status_code}",
                )

            data = resp.json().get("data", {})
            tweet_id = data.get("id", "")
            if i == 0:
                first_tweet_id = tweet_id
            previous_id = tweet_id

            # Delay between tweets to avoid rate limits
            if i < len(parts) - 1:
                await asyncio.sleep(THREAD_DELAY_MS / 1000)

        return PublishResult(
            success=True,
            platform="twitter",
            platform_post_id=first_tweet_id,
            platform_post_url=f"https://twitter.com/i/status/{first_tweet_id}",
            published_at=datetime.now(timezone.utc),
            metadata={"thread_length": len(parts)},
        )

    @staticmethod
    def _split_thread(content: str) -> list[str]:
        """Split content into thread parts using --- separator."""
        if "\n---\n" in content:
            return [part.strip() for part in content.split("\n---\n") if part.strip()]
        return [content]
